content/news/2019-12-03-mapstruct-and-quarkus.md

```diff
@@ -0,0 +1,220 @@
+We don’t like to expose internal objects, thus we create a new `PersonDto` that matches our API.
+
+{{< prettify java >}}
+public class PersonDto {
+  private String firstname;
+  private String surname;
+  
+  // getters, setters
+}
+{{< /prettify >}}
+
+So now, we have a `Person` and `PersonDto` that both have a similar structure. We are able to retrieve a `Person`, but want to return a `PersonDto`, so that’s the time where we finally add our MapStruct mapper!
+
+{{< prettify java >}}
+@Mapper(componentModel = "cdi")
+public interface PersonMapper {
+    @Mapping(target = "surname", source = "lastname")
+    PersonDto toResource(Person person);
+}
+{{< /prettify >}}
+
+The MapStruct annotation processor will generate the implementation for us, we just need to tell it that we would like to have a method that accepts a `Person` and returns a new `PersonDto`. A little trap: Our DTO has the property `surname`, so we add a mapping annotation to map `lastname` from `Person` to `surname` from `PersonDto`.
+
+**Very important:** The component model should be set to CDI, as this will allow us to easily inject the generated mapper implementation.
+
+_Side note: As you need to use CDI for all mappers it’s recommended to define a `@MapperConfig` and refer to it in all your mappers, this is the way I did it in the example on GitHub. But for simplicity I skipped it here._
+
+#### Endpoint
+As a last step we need to create our endpoint that will get the correct `Person` and finally returns the `PersonDto`.
+
+The `PersonResource` was already generated by Quarkus providing a `/person` endpoint that just returns a static text. We will remove this operation and exchange it with a new one that returns our person:
+
+{{< prettify java >}}
+@Path("/person")
+public class PersonResource {
+ 
+  @Inject
+  PersonService personService;
+ 
+  @Inject
+  PersonMapper personMapper;
+ 
+  @GET
+  @Produces(MediaType.APPLICATION_JSON)
+  public PersonDto loadPerson() {
+    return personMapper.toResource( personService.loadPerson() );
+  }
+}
+{{< /prettify >}}
+
+To be able to access our `PersonService` and `PersonMapper` we will inject both using CDI.
+
+The `loadPerson()` method will then just call our service to load the person, throw it into our mapper and return the DTO.
+
+That’s it!
+
+#### Start the application
+If you would like to run the application you just need to hit `mvn compile quarkus:dev`. This will compile everything (which includes mapper generation) and starts a local server running the application.
+
+After starting the application you should see a log output similar to the following one:
+
+{{< prettify >}}
+Quarkus 1.0.1.Final started in 1.448s. Listening on: http://0.0.0.0:8080
+Profile dev activated. Live Coding activated.
+Installed features: [cdi, resteasy, resteasy-jsonb]
+{{< /prettify >}}
+
+You can now open [http://localhost:8080/person](http://localhost:8080/person) in your browser and should see a JSON representating our person.
+
```

Comment:
If it works with MapStruct, I think it would be worth demoing the dev mode e.g. do some live changes and refresh.

Also it's not fair comparing dev mode to native image :). If you want to compare, build the runner jar, it should be a bit faster, I think. :)